onnxruntime/python/tools/automl/data_frame_tool.py

```diff
@@ -0,0 +1,137 @@
+        """
+        The function does the heavy lifting for _get_input_feeds()
+
+        :param df: See :class:`pandas.DataFrame`. 
+        :param input_metas: a list of name/type pairs
+        :require is a boolean. If True this helper throws on a missing input.
+        
+        """
+        feeds = {}
+        # Process mandadory inputs. Raise an error if anything is not present
+        for input_meta in input_metas:
+            shape = [dim if dim else 1 for dim in input_meta.shape]
+            # We fully expect all the types are in the above dictionary
+            assert input_meta.type in types_dict
+            print("Processing input: " + input_meta.name)
+            if input_meta.name in df.columns:
+                expected_type = types_dict[input_meta.type]
+                # float16 and bool will always require exact match
+                # We attempt to convert any type to a string if it is required.
+                # With strings we always want to put this into a flat array, cast to np.object and then reshape as object
+                if input_meta.type == 'tensor(string)':
+                    #print('Col: {} processed as string type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                elif expected_type == df[input_meta.name].dtype: # If there is an exact match we take as is
+                    #print('Col: {} processed exact match type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                elif expected_type in ort_float_set and str(df[input_meta.name].dtype) in pd_float_set:
+                    #print('Col: {} processed as floating type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                elif expected_type in ort_int_set and str(df[input_meta.name].dtype) in pd_int_set:
+                    #print('Col: {} processed as integer type: {} '.format(input_meta.name, df[input_meta.name].dtype))
+                    feeds[input_meta.name] = np.array([df[input_meta.name][0]]).astype(expected_type).reshape(shape)
+                else:
+                    raise TypeError("Input {} expected to be of type: {} got {} ".format(
+                                input_meta.name, expected_type, df[input_meta.name].dtype))
+            elif require:
+                raise RuntimeError("This model requires input {} of type {} but it is not found in the DataFrame".format(
+                               input_meta.name, types_dict[input_meta.type]))
+        return feeds
+
+
+    def _get_input_feeds(self, df, sess):
+        "Return a dictionary of input_name : a typed and shaped np.array of values"
```

Comment:
This comment should probably be the first line in the docstring on the next line.